### DISCUSSION TITLE
minio client API

### DISCUSSION THREAD

**Discussion by JayaChandra03** [minio/minio#15093] [Q&A] [answered]
can minio client API (mc) communicate with zenko s3 compatible server except that admin APIs?

**Comment by harshavardhana** [comment, marked as answer]
@JayaChandra03 we do not know how S3 compatible Zenko is ? - we did some similar testing with GCP's S3 compatibility claim and here are some results.. 

| API | AWS S3 | MinIO | GCS (S3 compatibility)|
| :-------------| ------------------- | -------------------  | ------------------- |
| [ListObjectVersions](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectVersions.html) | :heavy_check_mark: | :heavy_check_mark: | :heavy_exclamation_mark: |

GCS fails in this manner while they succeed on AWS S3 and MinIO fine.
```
~ mc ls --versions gcs/harshavardhana
mc: <ERROR> Unable to list folder. unrecognized option:Marker
```

| API | AWS S3 | MinIO | GCS (S3 compatibility)|
| :-------------| ------------------- | -------------------  | ------------------- |
| [GetBucketEncryption](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketEncryption.html) | :heavy_check_mark: | :heavy_check_mark: | :heavy_exclamation_mark: |


GCS fails in this manner while they succeed on AWS S3 and MinIO fine.
```
~  mc encrypt info gcs/harshavardhana
mc: <ERROR> Unable to get encryption info: 503 Service Unavailable.
```

| API | AWS S3 | MinIO | GCS (S3 compatibility)|
| :-------------| ------------------- | -------------------  | ------------------- |
| [PutBucketLifecycle](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketLifecycleConfiguration.html) | :heavy_check_mark: | :heavy_check_mark: | :heavy_exclamation_mark: |

GCS fails in this manner while they succeed on AWS S3 and MinIO fine.
```
~ mc ilm add gcs/harshavardhana --expiry-days 365
mc: <ERROR> Unable to set new lifecycle rules. The XML you provided was not well-formed or did not validate against our published schema.
```

| API | AWS S3 | MinIO | GCS (S3 compatibility)|
| :-------------| ------------------- | -------------------  | ------------------- |
| [SelectObjectContent](https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html) | :heavy_check_mark: | :heavy_check_mark: | :heavy_exclamation_mark: |

GCS fails in this manner while they succeed on AWS S3 and MinIO fine.
```
~ mc sql gcs/harshavardhana/hosts.csv
mc: <ERROR> Unable to run sql Invalid argument.
```

| API | AWS S3 | MinIO | GCS (S3 compatibility)|
| :-------------| ------------------- | -------------------  | ------------------- |
| [MultiObjectDelete](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html) | :heavy_check_mark: | :heavy_check_mark: | :heavy_exclamation_mark: |

GCS fails in this manner while they succeed on AWS S3 and MinIO fine.
```
~ mc rm gcs/harshavardhana/hosts
Removing `gcs/harshavardhana/hosts`.
mc: <ERROR> Failed to remove `gcs/harshavardhana/hosts`. Invalid argument.
```


| API | AWS S3 | MinIO | GCS (S3 compatibility)|
| :-------------| ------------------- | -------------------  | ------------------- |
| [PutObjectTags](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObjectTagging.html) | :heavy_check_mark: | :heavy_check_mark: | :heavy_exclamation_mark: |

GCS fails in this manner while they succeed on AWS S3 and MinIO fine.
```
~ mc tag set gcs/harshavardhana/hosts "key1=value1&key2=value2&key3=value3"
mc: <ERROR> Failed to set tags for https://storage.googleapis.com/harshavardhana/hosts: 503 Service Unavailable.
```

### Summary
AWS S3 API Compatibility is quite limited, since few basic APIs do not work well. Other APIs that are not mentioned are not part of GCS S3 compatibility layer implementation are 

- [Multipart Uploads](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html).
- [Multiple Destination Bucket Replication](https://aws.amazon.com/blogs/aws/new-amazon-s3-replication-adds-support-for-multiple-destination-buckets/) 
- [Event Notifications](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html)
- [Object Locking and Legal Hold](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock-overview.html)
